core/src/main/java/org/infinispan/encoding/DataConversion.java

```diff
@@ -120,9 +120,12 @@ public void overrideWrapper(Class<? extends Wrapper> newWrapper, ComponentRegist
    /**
     * Obtain the configured {@link MediaType} for this instance, or assume sensible defaults.
     */
-   private MediaType getStorageMediaType(Configuration configuration, boolean embeddedMode, boolean internalCache) {
+   private MediaType getStorageMediaType(Configuration configuration, boolean embeddedMode, boolean internalCache, PersistenceMarshaller persistenceMarshaller) {
       EncodingConfiguration encodingConfiguration = configuration.encoding();
       ContentTypeConfiguration contentTypeConfiguration = isKey ? encodingConfiguration.keyDataType() : encodingConfiguration.valueDataType();
+      Marshaller userMarshaller = persistenceMarshaller.getUserMarshaller();
+      MediaType mediaType = userMarshaller.mediaType();
```

Comment:
Could you clarify? The DataConversion is initialized when the cache is initialized, it's a component. Also the information about storage type, request type, transcoder, etc is part of the DataConversion

Reply:
I meant the user marshaller and the storage type must be in sync. I don't see any point in allowing a cache to be configured with both

```xml
<value media-type="application/x-protostream"/>
```
and

```xml
<serialization marshaller="org.infinispan.jboss.marshalling.core.JBossUserMarshaller"/>
```

>  ~it's a component~.

Exactly. There is a `DataConversion` component, but most of the `injectDependencies()` logic is skipped for that instance because the same class is trying to be 2 things at the same time.

Reply:
The pair ```<value media-type="application/x-protostream"/>``` and ```<serialization marshaller="org.infinispan.jboss.marshalling.core.JBossUserMarshaller"/>``` is possible. The later is cache-manager scoped, the former cache scope config. 

It means the cache will store protobuf and a transcoder that is able to convert between ```application/x-jboss-marshaller``` <=> ```application/x-java-object``` is registered. This transcoder can be used whenever another cache needs that conversion

Reply:
Good point, I was thinking of the user marshaller as a purely "to/from storage" thing, but it's obvious now that it could also be a "to/from request" thing.

However, `PersistenceManagerImpl.createUserMarshaller()` ignores the cache's storage media type, so `storageMediaType` will always be set to the user marshaller's media type.

And it also made me realize that the `<value media-type="X"/>` configuration serves a dual purpose. It was even more complicated before, but with your changes IIUC
* With `<binary/>`, the storage type is X, and the default request type is `application/x-java-object`.
* With `<object/>`, the storage type is `application/x-java-object`, and the default request type is X.

Reply:
Going back to your initial point, ```injectDependencies()``` should be always called when the cache is created, otherwise I fear for the transcoding stuff :smile:

An alternative to what was done here, that aligns with your suggestion, is to move this logic to configuration time, more precisely, if a cache is heap (OBJECT) and don't have a MediaType configured, let's set the key/value mediatype to the "user marshaller". I will try doing that.

Reply:
```
<value media-type="X"/>
```
This has a single purpose, define the format of the values in the cache. This is configuration and cannot be changed at runtime. If not specified, a best guess will be made, e.g., indexed server caches will be protobuf otherwise there's no way of having query to work.

The request type will be app/x-java-object by default, and it's changeable at runtime, that's how the server deals with multiple request formats in REST, for example. 
I will rather say what is planned for the BINARY/OFF_HEAP/OBJECT trio after the last PR which is https://issues.redhat.com/browse/ISPN-11424:
* BINARY and OBJECT disappears
* OFF_HEAP is a flag on/off to specify if data will be stored in heap or not
* if off_heap=true for a cache, then we check the media type: it it's not byte[] based format, we throw an exception. 

So we're left only with the mediatype and a flag heap/off-heap

Reply:
@danberindei WRT:
> IMO if the cache has a user marshaller and the storage type is not OBJECT, then the cache should already be initialized with the user marshaller's media type as the storage type

I was not able to set the MediaType of the cache at configuration time, since the user marshaller is initialized at runtime and doesn't necessary follow the ```<serialization>``` configuration.

WRT:
> But please have another look at the user marshaller and whether its media type can override the cache's media type with the current code

When the ```DataContainer``` is initialized, it gives priority to the configured MediaType:

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/encoding/DataConversion.java#L126-L128

Reply:
> `<value media-type="X"/>` [...] has a single purpose, define the format of the values in the cache.

Sorry you had to spell that out again, I read "user marshaller can be used both ways" and I jumped to "content type configuration can also be used both ways". I guess I couldn't accept that the encoding configuration would be simply ignored for caches with OBJECT storage :)

> When the DataConversion is initialized, it gives priority to the configured MediaType:

I missed that too!

> I was not able to set the MediaType of the cache at configuration time

If you mean during `ConfigurationBuilder.build()` then I agree, we can't query the user marshaller at that time.

My thinking was that the storage type could be computed in `InternalCacheFactory.buildEncodingCache()`, so it doesn't have to be recomputed every time the `DataConversion` instance is sent to another node.

Reply:
Let me try that

Reply:
It didn't work: during ```InternalCacheFactory.buildEncodingCache()``` the persistence marshaller is ```null``` in the registry

Reply:
Yeah, the component registry is not yet bootstrapped there. But the persistence marshaller is a global component, so it doesn't need the cache component registry.

Here's how I would like it to work: https://github.com/infinispan/infinispan/commit/c864043facbbc0d503336959247788affebb6bfa

Let me know if you want to include my suggestions here, otherwise I can integrate your PR as is and open a separate PR.

Reply:
Cool, let's separate from this PR please